File: examples/clj/test/com/rpl/agent/basic/keyvalue_store_agent_test.clj
(ns com.rpl.agent.basic.keyvalue-store-agent-test
  (:require
   [clojure.test :refer [deftest testing is]]
   [com.rpl.agent-o-rama :as aor]
   [com.rpl.rama :as rama]
   [com.rpl.rama.test :as rtest]
   [com.rpl.agent.basic.keyvalue-store-agent :refer [KeyValueStoreModule]]))

(deftest keyvalue-store-agent-test
  (System/gc)
  (testing "KeyValueStoreAgent example produces expected results"
    (with-open [ipc (rtest/create-ipc)]
      (rtest/launch-module! ipc KeyValueStoreModule {:tasks 1 :threads 1})

      (let [manager (aor/agent-manager ipc
                                       (rama/get-module-name
                                        KeyValueStoreModule))
            agent   (aor/agent-client manager "KeyValueStoreAgent")]

        (testing "set operation stores value"
          (let [result (aor/agent-invoke agent
                                         {:counter-name "test-counter"
                                          :operation    :set
                                          :value        42})]
            (is (= :set (:action result)))
            (is (= "test-counter" (:counter result)))
            (is (= 42 (:value result)))))

        (testing "get operation retrieves stored value"
          (let [result (aor/agent-invoke agent
                                         {:counter-name "test-counter"
                                          :operation    :get})]
            (is (= :get (:action result)))
            (is (= "test-counter" (:counter result)))
            (is (= 42 (:value result)))))

        (testing "increment operation increases value"
          (let [result (aor/agent-invoke agent
                                         {:counter-name "test-counter"
                                          :operation    :increment})]
            (is (= :increment (:action result)))
            (is (= "test-counter" (:counter result)))
            (is (= 42 (:previous-value result)))
            (is (= 43 (:new-value result)))))

        (testing "update operation adds to existing value"
          (let [result (aor/agent-invoke agent
                                         {:counter-name "test-counter"
                                          :operation    :update
                                          :value        7})]
            (is (= :update (:action result)))
            (is (= "test-counter" (:counter result)))
            (is (= 43 (:previous-value result)))
            (is (= 7 (:added-value result)))
            (is (= 50 (:new-value result)))))

        (testing "increment on non-existent counter starts at 0"
          (let [result (aor/agent-invoke agent
                                         {:counter-name "new-counter"
                                          :operation    :increment})]
            (is (= :increment (:action result)))
            (is (= "new-counter" (:counter result)))
            (is (= 0 (:previous-value result)))
            (is (= 1 (:new-value result)))))))))

File: examples/java/src/test/java/com/rpl/agent/basic/KeyValueStoreAgentTest.java
package com.rpl.agent.basic;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

import com.rpl.agentorama.AgentClient;
import com.rpl.agentorama.AgentManager;
import com.rpl.rama.test.InProcessCluster;
import com.rpl.rama.test.LaunchConfig;
import java.util.Map;
import org.junit.Test;

/**
 * Test class for KeyValueStoreAgent demonstrating persistent state management.
 *
 * <p>This test demonstrates:
 *
 * <ul>
 *   <li>declareKeyValueStore: Creating persistent key-value storage
 *   <li>getStore: Accessing stores from agent nodes
 *   <li>Store operations: get, put, update for persistent state
 *   <li>HashMap usage for request and response data structures
 * </ul>
 */
public class KeyValueStoreAgentTest {

  @Test
  public void testKeyValueStoreAgent() throws Exception {
    // Tests basic key-value store operations using HashMap
    try (InProcessCluster ipc = InProcessCluster.create()) {
      // Deploy the agent module
      KeyValueStoreAgent.KeyValueStoreModule module = new KeyValueStoreAgent.KeyValueStoreModule();
      ipc.launchModule(module, new LaunchConfig(1, 1));

      // Get agent manager and client
      String moduleName = module.getModuleName();
      AgentManager manager = AgentManager.create(ipc, moduleName);
      AgentClient agent = manager.getAgentClient("KeyValueStoreAgent");

      // Test SET operation
      Map<String, Object> setRequest =
          KeyValueStoreAgent.createCounterRequest(
              "test-counter", KeyValueStoreAgent.Operation.SET, 42L);
      Map<String, Object> setResult = (Map<String, Object>) agent.invoke(setRequest);

      assertNotNull("Set result should not be null", setResult);
      assertEquals("Action should be set", "set", setResult.get("action"));
      assertEquals("Counter name should match", "test-counter", setResult.get("counter"));
      assertEquals("Value should be 42", 42L, setResult.get("value"));

      // Test GET operation
      Map<String, Object> getRequest =
          KeyValueStoreAgent.createCounterRequest(
              "test-counter", KeyValueStoreAgent.Operation.GET, null);
      Map<String, Object> getResult = (Map<String, Object>) agent.invoke(getRequest);

      assertNotNull("Get result should not be null", getResult);
      assertEquals("Action should be get", "get", getResult.get("action"));
      assertEquals("Counter name should match", "test-counter", getResult.get("counter"));
      assertEquals("Value should be 42", 42L, getResult.get("value"));
    }
  }
}
